### DISCUSSION TITLE
NoSuchModuleError: SQLAlchemy.dialects

### DISCUSSION THREAD

**Discussion by EurasiaHewi** [sqlalchemy/sqlalchemy#10576] [Usage Questions]
Hey there,

I hope you might help with implementing the following project:

https://github.com/Nukesor/ultimate-poll-bot

I am no expert in coding, especially not in handling databases, so maybe I am doing something completely dumb here...

Within the main file a database is set up:

```py
@cli.command()
def initdb(exist_ok: bool = False, drop_existing: bool = False):

    db_url = engine.url
    typer.echo(f"Using database at {db_url}")

    if database_exists(db_url):
        if drop_existing:
            with wrap_echo("Dropping database"):
                drop_database(db_url)
        elif not exist_ok:
            typer.echo(
                "Database already exists, aborting.\n"
                "Use --exist-ok if you are sure the database is uninitialized and contains no data.\n"
                "Use --drop-existing if you want to recreate it.",
                err=True,
            )
            return
```
            
Here, it accesses the following code (saved in a extra script) to set up the engine:

```py
from typing import cast

from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, scoped_session, sessionmaker

from pollbot.config import config

engine = create_engine(
    config["database"]["sql_uri"],
    pool_size=config["database"]["connection_count"],
    max_overflow=config["database"]["overflow_count"],
    echo=False,
)
```

The following error is produced:

```
  File ~\anaconda3\Lib\site-packages\sqlalchemy\util\langhelpers.py:367 in load
    raise exc.NoSuchModuleError(

NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:postgressql
```

A search in the web mostly leads to the hint, that one need to change postgres into postgressql to prevent this error. However, the places where this order shows up, do not use postgres orders:

```
[database]
sql_uri = "postgressql://localhost/pollbot"
connection_count = 10
overflow_count = 5
```

....

```
"database": {
        "sql_uri": "postgressql://pollbot:localhost/pollbot",
        "connection_count": 20,
        "overflow_count": 10,
    }
```
    
 Any idea?
 
 Greetings, Hewi
 
 
 PS: The head of the first code is not portrayed as code, don't know why. But it belongs to the code there....

**Comment by EurasiaHewi** [comment]
The files used to set up the databse and define the url are (as far as I can see...):

main.py
config.py (in pollbot directory)
ultimate_pollbot.toml

**Comment by gordthompson** [comment]
Use `postgresql://`, not `postgressql://`

**Reply by EurasiaHewi** [reply]
This database setup is a pain:

```py
from typing import cast

from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, scoped_session, sessionmaker

from pollbot.config import config

engine = create_engine(
    config["database"]["sql_uri"],
    pool_size=config["database"]["connection_count"],
    max_overflow=config["database"]["overflow_count"],
    echo=False,
)
base = declarative_base(bind=engine)
```


Eror:

TypeError: declarative_base() got an unexpected keyword argument 'bind'

Maybe you see the problem here in a flash as well... :D

**Reply by gordthompson** [reply]
As per

https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.declarative_base

> **Deprecated since version 1.4:** The “bind” argument to declarative_base is deprecated and will be removed in SQLAlchemy 2.0.

**Reply by EurasiaHewi** [reply]
Ah, I see. Is there a workaround or should I just let it free? This error vanished by simply dropping any argument in the brackets. Is its function deviating completely now, so does it makes sense to try it wouth the bind parameter?

**Reply by gordthompson** [reply]
Just omit the `bind=` argument and associate ("bind") the engine to your Session

```python
with Session(engine) as session:
    # do stuff
```

ref:

https://docs.sqlalchemy.org/en/20/tutorial/dbapi_transactions.html#tutorial-executing-orm-session

**Reply by EurasiaHewi** [reply]
This is the complete db.py code:

```"""Helper class to get a database engine and to get a session."""
from typing import cast

from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, scoped_session, sessionmaker

from pollbot.config import config

engine = create_engine(
    config["database"]["sql_uri"],
    pool_size=config["database"]["connection_count"],
    max_overflow=config["database"]["overflow_count"],
    echo=False,
)
base = declarative_base(bind=engine)


def get_session(connection: None = None) -> Session:
    """Get a new db session."""
    session = scoped_session(sessionmaker(bind=engine))
    return cast(Session, session)```

**Reply by EurasiaHewi** [reply]
Unfortunately I really don't have a clue, what the code is doing here. I have read the reference, but it is full of other terms I lack. I can't use it with my current level of database knowledge.

For me it looks like any database requires an engine (no idea, what it is doing). The regarded code "binds" the engine to the base (don't know what it is doing either). 

So where should I run this association, inside the engine creation, or when the function is called in the main file? What should I write here: #do stuff


Thank your for the fast responses!

**Reply by CaselIT** [reply]
My suggestion would to go trough the tutorial, that will explain the basis of how to use sqlalchemy https://docs.sqlalchemy.org/en/20/tutorial/index.html

quickstart may also be useful https://docs.sqlalchemy.org/en/20/orm/quickstart.html

**Reply by EurasiaHewi** [reply]
I read some relevant passages, the tutorial is written well, but I don't get what to do. I think this is essentially what I want to reproduce for my case:

```>>> from sqlalchemy.orm import Session

>>> stmt = text("SELECT x, y FROM some_table WHERE y > :y ORDER BY x, y")
>>> with Session(engine) as session:
...     result = session.execute(stmt, {"y": 6})
...     for row in result:
...         print(f"x: {row.x}  y: {row.y}")

**Reply by EurasiaHewi** [reply]
I don't have the time to really learn to set up a database. I need a international working poll bot, to process poll data (results, ect..). It appears like this needs a database. But I don't know which data in particular are stored, or how. I guess the code above stores data in "result". But how this can be integrated in the existing code is not clear to me.

Thank you though, maybe I get it tomorrow.. :D

**Comment by EurasiaHewi** [comment]
Oh boy, several hours of search... And than it is the spelling...

Many thanks, solved it

**Comment by CaselIT** [comment]
> PS: The head of the first code is not portrayed as code, don't know why. But it belongs to the code there....

you need to use 3 ``` ` ``` with blocks of code

**Reply by EurasiaHewi** [reply]
Thank you!

**Comment by gordthompson** [comment]
My apologies. I now realize that you aren't asking about changing your own code, you just want to get somebody else's code running and that code has not been updated in a long time. In that case you might be able to install an old version of SQLAlchemy via

```
pip install "SQLAlchemy<1.4"
```

and that may be enough to get it working. (Although some other dependencies may also need to be pinned to older versions.)

**Reply by EurasiaHewi** [reply]
Thank you, I will try it !
